aten/src/ATen/templates/TensorBody.h

```diff
@@ -394,13 +447,174 @@ class CAFFE2_API Tensor {
     return impl_->requires_grad();
   }
 
+  // TODO: There is something bad about this: const-correctness on Tensor
+  // is not supposed to matter, but here we are looking at the constness
+  // of Tensor to decide whether or not to return a mutable reference...
   Tensor& grad() {
     return impl_->grad();
   }
   const Tensor& grad() const {
     return impl_->grad();
   }
 
+  /// NOTE: This is similar to the legacy `.data()` function on `Variable`, and is intended
+  /// to be used from functions that need to access the `Variable`'s equivalent `Tensor`
+  /// (i.e. `Tensor` that shares the same storage and tensor metadata with the `Variable`).
+  ///
+  /// One notable difference with the legacy `.data()` function is that changes to the
+  /// returned `Tensor`'s tensor metadata (e.g. sizes / strides / storage / storage_offset)
+  /// will not update the original `Variable`, due to the fact that this function
+  /// shallow-copies the `Variable`'s underlying TensorImpl.
+  Tensor tensor_data() const noexcept {
+    auto self_impl_copy = impl_->shallow_copy_and_detach(
+      /*version_counter=*/impl_->version_counter(),
+      /*allow_tensor_metadata_change=*/impl_->allow_tensor_metadata_change());
+    return Tensor(self_impl_copy);
+  }
+
+  /// NOTE: `var.variable_data()` in C++ has the same semantics as `tensor.data`
```

Comment:
Does that mean that `var.variable_data()` is the same as `var.detach()`?

Reply:
cc @yf225 this is just preexisting

Reply:
I think the only difference between `var.variable_data()` (aka. `tensor.data` in Python) and `var.detach()` (aka. `tensor.detach()` in Python) is that the former doesn't share version counter, but the latter does.